File: apps/docker-server/cmd/renewlet/custom_head_html.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"strings"
	"unicode/utf8"

	"golang.org/x/net/html"
	"golang.org/x/net/html/atom"
)

const (
	customHeadHTMLEnvName  = "RENEWLET_CUSTOM_HEAD_HTML"
	customHeadHTMLMaxBytes = 64 * 1024
)

type customHeadHTMLConfig struct {
	Markup string
}

func (config customHeadHTMLConfig) Enabled() bool {
	return config.Markup != ""
}

func customHeadHTMLFromEnv() (customHeadHTMLConfig, error) {
	return parseCustomHeadHTML(os.Getenv(customHeadHTMLEnvName))
}

func parseCustomHeadHTML(raw string) (customHeadHTMLConfig, error) {
	if len(raw) > customHeadHTMLMaxBytes {
		return customHeadHTMLConfig{}, fmt.Errorf("%s exceeds the 64 KiB UTF-8 limit", customHeadHTMLEnvName)
	}
	if !utf8.ValidString(raw) {
		return customHeadHTMLConfig{}, fmt.Errorf("%s must be valid UTF-8", customHeadHTMLEnvName)
	}
	if strings.TrimSpace(raw) == "" {
		return customHeadHTMLConfig{}, nil
	}

	// HTML tree builder 会按规范恢复错误标签，因此“能解析”不代表源码仍被约束在 head 内；先拦截顶层文档边界，再用尾部哨兵确认片段没有吞掉后续宿主结构。
	if err := rejectDocumentEscapeTokens(raw); err != nil {
		return customHeadHTMLConfig{}, err
	}
	marker := customHeadHTMLMarker(raw)
	nodes, err := html.ParseFragment(
		strings.NewReader(raw+"<!--"+marker+"-->"),
		&html.Node{Type: html.ElementNode, DataAtom: atom.Head, Data: "head"},
	)
	if err != nil {
		return customHeadHTMLConfig{}, fmt.Errorf("parse %s: %w", customHeadHTMLEnvName, err)
	}
	if len(nodes) == 0 || nodes[len(nodes)-1].Type != html.CommentNode || nodes[len(nodes)-1].Data != marker {
		return customHeadHTMLConfig{}, fmt.Errorf("%s must be a complete head fragment", customHeadHTMLEnvName)
	}

	allowedElements := map[string]bool{
		"base": true, "link": true, "meta": true, "noscript": true,
		"script": true, "style": true, "template": true, "title": true,
	}
	for _, node := range nodes[:len(nodes)-1] {
		switch node.Type {
		case html.TextNode:
			if strings.TrimSpace(node.Data) != "" {
				return customHeadHTMLConfig{}, fmt.Errorf("%s must not contain non-whitespace top-level text", customHeadHTMLEnvName)
			}
		case html.CommentNode:
		case html.ElementNode:
			if node.Namespace != "" || !allowedElements[node.Data] {
				return customHeadHTMLConfig{}, fmt.Errorf("%s contains <%s>, which is not valid at the top level of a head fragment", customHeadHTMLEnvName, node.Data)
			}
		default:
			return customHeadHTMLConfig{}, fmt.Errorf("%s contains unsupported top-level markup", customHeadHTMLEnvName)
		}
	}

	// 该变量由实例部署者控制并拥有页面同源代码权限；这里只守住 head 结构，不清洗属性、重写脚本或推断外部资源。
	return customHeadHTMLConfig{Markup: raw}, nil
}

func rejectDocumentEscapeTokens(raw string) error {
	tokenizer := html.NewTokenizer(strings.NewReader(raw))
	// 这里只维护源码层的开放元素栈，用于区分顶层 document boundary escape 与 script/template 内部内容；它不是第二套 HTML 校验器。
	openElements := []string{}
	for {
		tokenType := tokenizer.Next()
		switch tokenType {
		case html.ErrorToken:
			if err := tokenizer.Err(); err != nil && !errors.Is(err, io.EOF) {
				return fmt.Errorf("parse %s: %w", customHeadHTMLEnvName, err)
			}
			return nil
		case html.DoctypeToken:
			return fmt.Errorf("%s must not contain a doctype", customHeadHTMLEnvName)
		case html.StartTagToken, html.SelfClosingTagToken:
			name, _ := tokenizer.TagName()
			tagName := strings.ToLower(string(name))
			if len(openElements) == 0 && isHTMLDocumentBoundaryTag(tagName) {
				return fmt.Errorf("%s must not contain head, html, or body tags", customHeadHTMLEnvName)
			}
			if tokenType == html.StartTagToken && !isVoidHTMLElement(tagName) {
				openElements = append(openElements, tagName)
			}
		case html.EndTagToken:
			name, _ := tokenizer.TagName()
			tagName := strings.ToLower(string(name))
			if len(openElements) == 0 && isHTMLDocumentBoundaryTag(tagName) {
				return fmt.Errorf("%s must not contain head, html, or body tags", customHeadHTMLEnvName)
			}
			matched := false
			for index := len(openElements) - 1; index >= 0; index-- {
				if openElements[index] == tagName {
					openElements = openElements[:index]
					matched = true
					break
				}
			}
			if !matched && len(openElements) == 0 {
				return fmt.Errorf("%s contains an unmatched top-level closing tag", customHeadHTMLEnvName)
			}
		}
	}
}

func isHTMLDocumentBoundaryTag(tagName string) bool {
	return tagName == "head" || tagName == "html" || tagName == "body"
}

func isVoidHTMLElement(tagName string) bool {
	switch tagName {
	case "area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "param", "source", "track", "wbr":
		return true
	default:
		return false
	}
}

func customHeadHTMLMarker(raw string) string {
	marker := "renewlet-custom-head-html-end"
	for strings.Contains(raw, marker) {
		marker += "-marker"
	}
	return marker
}

type customHeadHTMLFS struct {
	fs.FS
	index     []byte
	indexInfo fs.FileInfo
}

func prepareCustomHeadHTMLFS(staticFS fs.FS, config customHeadHTMLConfig) (fs.FS, error) {
	if !config.Enabled() {
		return staticFS, nil
	}
	content, err := fs.ReadFile(staticFS, "index.html")
	if err != nil {
		return nil, fmt.Errorf("read embedded index.html: %w", err)
	}
	info, err := fs.Stat(staticFS, "index.html")
	if err != nil {
		return nil, fmt.Errorf("stat embedded index.html: %w", err)
	}
	injected, err := injectCustomHeadHTML(content, config)
	if err != nil {
		return nil, err
	}
	// 启动期只物化 index.html；其余资产继续委托给 embedded FS，避免破坏 hash 缓存、SPA fallback 与预压缩协商。
	return customHeadHTMLFS{FS: staticFS, index: injected, indexInfo: info}, nil
}

func (fsys customHeadHTMLFS) Open(name string) (fs.File, error) {
	if name != "index.html" {
		return fsys.FS.Open(name)
	}
	return &staticMemoryFile{
		Reader: bytes.NewReader(fsys.index),
		info:   staticFileInfo{FileInfo: fsys.indexInfo, size: int64(len(fsys.index))},
	}, nil
}

type staticMemoryFile struct {
	*bytes.Reader
	info fs.FileInfo
}

func (file *staticMemoryFile) Stat() (fs.FileInfo, error) {
	return file.info, nil
}

func (file *staticMemoryFile) Close() error {
	return nil
}

type staticFileInfo struct {
	fs.FileInfo
	size int64
}

func (info staticFileInfo) Size() int64 {
	// apis.Static 依据 FileInfo 生成 Content-Length；注入后的内存文件不能沿用 embedded index.html 的原始大小。
	return info.size
}

func injectCustomHeadHTML(content []byte, config customHeadHTMLConfig) ([]byte, error) {
	if !config.Enabled() {
		return content, nil
	}
	// tokenizer 只定位宿主的显式 </head>，按原始字节偏移注入，避免 DOM 序列化改写页面或部署者片段。
	closeOffset, err := explicitHeadCloseOffset(content)
	if err != nil {
		return nil, err
	}
	var output bytes.Buffer
	output.Grow(len(content) + len(config.Markup) + 2)
	output.Write(content[:closeOffset])
	output.WriteByte('\n')
	output.WriteString(config.Markup)
	output.WriteByte('\n')
	output.Write(content[closeOffset:])
	return output.Bytes(), nil
}

func explicitHeadCloseOffset(content []byte) (int, error) {
	tokenizer := html.NewTokenizer(bytes.NewReader(content))
	offset := 0
	foundHead := false
	for {
		tokenType := tokenizer.Next()
		rawToken := tokenizer.Raw()
		startOffset := offset
		offset += len(rawToken)
		switch tokenType {
		case html.ErrorToken:
			if err := tokenizer.Err(); err != nil && !errors.Is(err, io.EOF) {
				return 0, fmt.Errorf("parse embedded index.html: %w", err)
			}
			return 0, errors.New("embedded index.html must contain an explicit head element")
		case html.StartTagToken:
			name, _ := tokenizer.TagName()
			if bytes.EqualFold(name, []byte("head")) {
				foundHead = true
			}
		case html.EndTagToken:
			name, _ := tokenizer.TagName()
			if foundHead && bytes.EqualFold(name, []byte("head")) {
				return startOffset, nil
			}
		}
	}
}

File: apps/web/vite/custom-head-html.ts
import { defaultTreeAdapter, html as parse5HTML, parse, parseFragment } from "parse5";

export const customHeadHTMLEnvName = "RENEWLET_CUSTOM_HEAD_HTML";
export const customHeadHTMLMaxBytes = 64 * 1024;

const customHeadHTMLStartMarker = "<!-- renewlet-custom-head-html:start -->";
const customHeadHTMLEndMarker = "<!-- renewlet-custom-head-html:end -->";
const allowedHeadElements = new Set(["base", "link", "meta", "noscript", "script", "style", "template", "title"]);

export type CustomHeadHTML = {
  markup: string;
};

type ParsedNode = {
  nodeName: string;
  tagName?: string;
  value?: string;
  data?: string;
  childNodes?: ParsedNode[];
  sourceCodeLocation?: {
    startOffset: number;
    endOffset: number;
    endTag?: { startOffset: number };
  };
};

export function parseCustomHeadHTML(raw: string | undefined): CustomHeadHTML | undefined {
  const markup = raw ?? "";
  if (!hasWellFormedUnicode(markup)) {
    throw new Error(`${customHeadHTMLEnvName} must be valid UTF-8`);
  }
  if (new TextEncoder().encode(markup).byteLength > customHeadHTMLMaxBytes) {
    throw new Error(`${customHeadHTMLEnvName} exceeds the 64 KiB UTF-8 limit`);
  }
  if (!markup.trim()) return undefined;

  // WHATWG tree builder 会恢复并重排错误标签；唯一尾部哨兵用于证明片段没有吞掉宿主页后续结构，不能把“parse5 未报错”当作合法性结论。
  const marker = customHeadHTMLMarker(markup);
  const source = `${markup}<!--${marker}-->`;
  const context = defaultTreeAdapter.createElement("head", parse5HTML.NS.HTML, []);
  const fragment = parseFragment(context, source, { sourceCodeLocationInfo: true });
  const nodes = fragment.childNodes as ParsedNode[];
  const endNode = nodes.at(-1);
  if (endNode?.nodeName !== "#comment" || endNode.data !== marker || endNode.sourceCodeLocation?.startOffset !== markup.length) {
    throw new Error(`${customHeadHTMLEnvName} must be a complete head fragment`);
  }

  // parse5 可能静默忽略或重定位源码 token；location coverage 要求每段非空白源码都映射到稳定的 head 顶层节点。
  let coveredOffset = 0;
  for (const node of nodes) {
    const location = node.sourceCodeLocation;
    if (!location || location.startOffset < coveredOffset) {
      throw new Error(`${customHeadHTMLEnvName} could not be mapped to a stable head fragment`);
    }
    if (source.slice(coveredOffset, location.startOffset).trim()) {
      throw new Error(`${customHeadHTMLEnvName} must not contain html or body escape markup`);
    }
    coveredOffset = location.endOffset;
  }
  if (coveredOffset !== source.length) {
    throw new Error(`${customHeadHTMLEnvName} could not be mapped to a complete head fragment`);
  }

  for (const node of nodes.slice(0, -1)) {
    if (node.nodeName === "#text") {
      if (node.value?.trim()) {
        throw new Error(`${customHeadHTMLEnvName} must not contain non-whitespace top-level text`);
      }
      continue;
    }
    if (node.nodeName === "#comment") continue;
    if (!node.tagName || !allowedHeadElements.has(node.tagName)) {
      throw new Error(`${customHeadHTMLEnvName} contains <${node.tagName ?? node.nodeName}>, which is not valid at the top level of a head fragment`);
    }
  }

  // 该变量由实例部署者控制并拥有页面同源代码权限；这里只守住 head 结构，不清洗属性、重写脚本或推断外部资源。
  return { markup };
}

export function transformCustomHeadHTML(documentHTML: string, customHeadHTML: CustomHeadHTML | undefined): string {
  if (!customHeadHTML) return documentHTML;
  // Vite 的同一 HTML hook 在 dev/build 或插件组合中可能重复执行；成对 marker 只负责幂等，不承担内容清洗或信任判断。
  const hasStartMarker = documentHTML.includes(customHeadHTMLStartMarker);
  const hasEndMarker = documentHTML.includes(customHeadHTMLEndMarker);
  if (hasStartMarker || hasEndMarker) {
    if (hasStartMarker && hasEndMarker) return documentHTML;
    throw new Error("Renewlet custom head HTML markers are incomplete.");
  }

  const document = parse(documentHTML, { sourceCodeLocationInfo: true }) as unknown as ParsedNode;
  const head = findElement(document, "head");
  const closeOffset = head?.sourceCodeLocation?.endTag?.startOffset;
  if (closeOffset === undefined) {
    throw new Error("Vite index.html must contain an explicit head element.");
  }

  // parse5 只定位宿主的显式 </head>；最终按源码偏移拼接，避免序列化部署者提供的脚本、属性和多标签片段。
  const injected = [customHeadHTMLStartMarker, customHeadHTML.markup, customHeadHTMLEndMarker].join("\n");
  return `${documentHTML.slice(0, closeOffset)}\n${injected}\n${documentHTML.slice(closeOffset)}`;
}

export function updateCustomHeadHTMLStaticHeaders(headers: string, customHeadHTML: CustomHeadHTML | undefined): string {
  if (!customHeadHTML) return headers;
  const cspLinePattern = /^(\s*Content-Security-Policy:\s*)(.+)$/m;
  if (!cspLinePattern.test(headers)) {
    throw new Error("Missing Content-Security-Policy in apps/web/dist/_headers.");
  }
  return headers.replace(cspLinePattern, `$1${trustedExtensionContentSecurityPolicy(true)}`);
}

export function trustedExtensionContentSecurityPolicy(httpsProduction: boolean): string {
  // 可信扩展态刻意不声明 default-src 与 fetch directives；任何 fallback 白名单都会重新限制动态脚本、XHR、图片和 frame。
  const directives = ["object-src 'none'", "base-uri 'self'", "frame-ancestors 'none'", "form-action 'self'"];
  if (httpsProduction) directives.push("upgrade-insecure-requests");
  return directives.join("; ");
}

function findElement(node: ParsedNode, tagName: string): ParsedNode | undefined {
  if (node.tagName === tagName) return node;
  for (const child of node.childNodes ?? []) {
    const match = findElement(child, tagName);
    if (match) return match;
  }
  return undefined;
}

function customHeadHTMLMarker(markup: string): string {
  let marker = "renewlet-custom-head-html-parse-end";
  while (markup.includes(marker)) marker += "-marker";
  return marker;
}

function hasWellFormedUnicode(value: string): boolean {
  // TextEncoder 会把孤立 UTF-16 surrogate 静默替换成 U+FFFD；先显式拒绝，才能让前后端共享“有效 UTF-8 输入”契约。
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return false;
    }
  }
  return true;
}
